flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphPartitionReleaseTest.java

```diff
@@ -76,52 +81,78 @@ public void testStrategyNotifiedOfFinishedVerticesAndResultsRespected() throws E
+		run(() -> {
+			final Execution sourceExecution = getCurrentExecution(sourceVertex, executionGraph);
+			executionGraph.updateState(new TaskExecutionState(executionGraph.getJobID(), sourceExecution.getAttemptId(), ExecutionState.FINISHED));
+			assertThat(releasedPartitions, empty());
+		});
+
+		run(() -> {
+			final Execution sourceExecution = getCurrentExecution(sourceVertex, executionGraph);
+			final Execution operatorExecution = getCurrentExecution(operatorVertex, executionGraph);
+			executionGraph.updateState(new TaskExecutionState(executionGraph.getJobID(), operatorExecution.getAttemptId(), ExecutionState.FINISHED));
+			assertThat(releasedPartitions, hasSize(1));
+			assertThat(releasedPartitions.remove(), equalTo(new ResultPartitionID(
+				sourceExecution.getVertex().getProducedPartitions().keySet().iterator().next(),
+				sourceExecution.getAttemptId())));
+		});
+
+		run(() -> {
+			final Execution operatorExecution = getCurrentExecution(operatorVertex, executionGraph);
+			final Execution sinkExecution = getCurrentExecution(sinkVertex, executionGraph);
+			executionGraph.updateState(new TaskExecutionState(executionGraph.getJobID(), sinkExecution.getAttemptId(), ExecutionState.FINISHED));
+
+			assertThat(releasedPartitions, hasSize(1));
+			assertThat(releasedPartitions.remove(), equalTo(new ResultPartitionID(
+				operatorExecution.getVertex().getProducedPartitions().keySet().iterator().next(),
+				operatorExecution.getAttemptId())));
+		});
 	}
 
-	private static Execution getExecution(final JobVertex jobVertex, final ExecutionGraph executionGraph) {
+	private static Execution getCurrentExecution(final JobVertex jobVertex, final ExecutionGraph executionGraph) {
 		return executionGraph.getJobVertex(jobVertex.getID()).getTaskVertices()[0].getCurrentExecutionAttempt();
 	}
 
 	private ExecutionGraph createExecutionGraph(final PartitionTracker partitionTracker, final JobVertex... vertices) throws Exception {
-		return ExecutionGraphBuilder.buildGraph(
+		final ExecutionGraph executionGraph = ExecutionGraphBuilder.buildGraph(
 			null,
 			new JobGraph(new JobID(), "test job", vertices),
 			new Configuration(),
-			TestingUtils.defaultExecutor(),
-			TestingUtils.defaultExecutor(),
+			scheduledExecutorService,
+			mainThreadExecutor,
 			new TestingSlotProvider(ignored -> CompletableFuture.completedFuture(new TestingLogicalSlot())),
 			ExecutionGraphPartitionReleaseTest.class.getClassLoader(),
 			new StandaloneCheckpointRecoveryFactory(),
 			AkkaUtils.getDefaultTimeout(),
-			new InfiniteDelayRestartStrategy(1),
+			new NoRestartStrategy(),
 			new UnregisteredMetricsGroup(),
 			VoidBlobWriter.getInstance(),
 			AkkaUtils.getDefaultTimeout(),
 			log,
 			NettyShuffleMaster.INSTANCE,
 			partitionTracker);
+
+		executionGraph.start(mainThreadExecutor);
+		run (executionGraph::scheduleForExecution);
+
+		return executionGraph;
+	}
+
+	private void run(RunnableWithException r) throws Exception {
```

Comment:
I think you can use `TestingComponentMainThreadExecutor` and `TestingComponentMainThreadExecutor.Resource`

Reply:
the resource does not provide access to the ScheduledExecutorService, which I need for the ExecutionGraph constructor.

Reply:
The `TestingComponentMainThreadExecutor` can be used though.